File: packages/sandboxd/src/process/control.rs
//! Control handles for runtime client process components.
//!
//! These methods expose restart, stop, readiness, and health operations without
//! leaking the process module's child/thread bookkeeping to higher-level
//! sandbox state code.

use std::sync::atomic::Ordering;
use std::sync::{Arc, Mutex};

use crate::cgroups::attach_pid_to_scope;
use crate::process::*;
use crate::supervision::SupervisedComponent;
use crate::time::{Clock, Sleeper};

impl CodexAppServerControlHandle {
    pub fn observation_handle(&self) -> &CodexAppServerObservationHandle {
        &self.managed_process.observation_handle
    }

    pub fn restart(&self, clock: &dyn Clock, sleeper: &dyn Sleeper) -> Result<(), String> {
        let _restart_guard = self
            .managed_process
            .restart_lock
            .lock()
            .expect("Codex app-server restart lock should not be poisoned");
        self.managed_process
            .restart_in_progress
            .store(true, Ordering::Relaxed);

        self.managed_process
            .supervisor_handle
            .mark_component_starting(SupervisedComponent::CodexAppServer);

        let mut current_process = RunningRuntimeClientProcess {
            spec: self.managed_process.spec.clone(),
            child: self.managed_process.child.clone(),
            output_capture: self.managed_process.output_capture.clone(),
            platform_scope: self.managed_process.platform_scope.clone(),
        };
        let _ = stop_runtime_client_process(&mut current_process, clock, sleeper);

        let (mut replacement_child, replacement_output_capture) =
            match spawn_runtime_client_child(&self.managed_process.spec) {
                Ok(replacement_child) => replacement_child,
                Err(error) => {
                    self.managed_process
                        .supervisor_handle
                        .mark_component_restarting(
                            SupervisedComponent::CodexAppServer,
                            error.clone(),
                        );
                    self.managed_process
                        .supervisor_handle
                        .emit_component_healthcheck_failed(
                            SupervisedComponent::CodexAppServer,
                            "restart_spawn_failed",
                            &error,
                            "process_liveness",
                            &[],
                        );
                    self.managed_process
                        .restart_in_progress
                        .store(false, Ordering::Relaxed);
                    return Err(error);
                }
            };
        let replacement_pid = replacement_child.id();
        if let Some(platform_scope) = &self.managed_process.platform_scope
            && let Err(error) = attach_pid_to_scope(&platform_scope.scope_paths, replacement_pid)
        {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            let error = error.to_string();
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::CodexAppServer, error.clone());
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        if let Some(platform_scope) = &self.managed_process.platform_scope
            && let Err(error) = platform_scope
                .registry
                .update_supervised_root_pid(&platform_scope.registry_key, replacement_pid)
        {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::CodexAppServer, error.clone());
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        if let Err(error) = wait_for_runtime_client_process_readiness_with(
            &self.managed_process.spec,
            &mut replacement_child,
            clock,
            sleeper,
        ) {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::CodexAppServer, error.clone());
            self.managed_process
                .supervisor_handle
                .emit_component_healthcheck_failed(
                    SupervisedComponent::CodexAppServer,
                    "restart_readiness_failed",
                    &error,
                    "readiness_ws",
                    &[],
                );
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        {
            let mut child = self
                .managed_process
                .child
                .lock()
                .expect("runtime client child lock should not be poisoned");
            *child = replacement_child;
        }
        update_codex_app_server_observation(
            &self.managed_process.observation_handle,
            &self.managed_process.spec,
            Some(replacement_pid),
            true,
            None,
        );
        self.managed_process
            .supervisor_handle
            .replace_component_details(
                SupervisedComponent::CodexAppServer,
                codex_app_server_details_with_status(
                    &self.managed_process.spec,
                    Some(replacement_pid),
                    None,
                    "Alive",
                    "Ready",
                ),
            );
        self.managed_process
            .supervisor_handle
            .mark_component_healthy(SupervisedComponent::CodexAppServer);
        self.managed_process
            .restart_in_progress
            .store(false, Ordering::Relaxed);
        Ok(())
    }
}

impl OpenCodeServerControlHandle {
    pub fn restart(&self, clock: &dyn Clock, sleeper: &dyn Sleeper) -> Result<(), String> {
        let _restart_guard = self
            .managed_process
            .restart_lock
            .lock()
            .map_err(|error| format!("OpenCode server restart lock was poisoned: {error}"))?;
        self.managed_process
            .restart_in_progress
            .store(true, Ordering::Relaxed);

        self.managed_process
            .supervisor_handle
            .mark_component_starting(SupervisedComponent::OpenCodeServer);

        let mut current_process = RunningRuntimeClientProcess {
            spec: self.managed_process.spec.clone(),
            child: self.managed_process.child.clone(),
            output_capture: self.managed_process.output_capture.clone(),
            platform_scope: self.managed_process.platform_scope.clone(),
        };
        let _ = stop_runtime_client_process(&mut current_process, clock, sleeper);

        let (mut replacement_child, replacement_output_capture) =
            match spawn_runtime_client_child(&self.managed_process.spec) {
                Ok(replacement_child) => replacement_child,
                Err(error) => {
                    self.managed_process
                        .supervisor_handle
                        .mark_component_restarting(
                            SupervisedComponent::OpenCodeServer,
                            error.clone(),
                        );
                    self.managed_process
                        .supervisor_handle
                        .emit_component_healthcheck_failed(
                            SupervisedComponent::OpenCodeServer,
                            "restart_spawn_failed",
                            &error,
                            "process_liveness",
                            &[],
                        );
                    self.managed_process
                        .restart_in_progress
                        .store(false, Ordering::Relaxed);
                    return Err(error);
                }
            };
        let replacement_pid = replacement_child.id();
        if let Some(platform_scope) = &self.managed_process.platform_scope
            && let Err(error) = attach_pid_to_scope(&platform_scope.scope_paths, replacement_pid)
        {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            let error = error.to_string();
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::OpenCodeServer, error.clone());
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        if let Some(platform_scope) = &self.managed_process.platform_scope
            && let Err(error) = platform_scope
                .registry
                .update_supervised_root_pid(&platform_scope.registry_key, replacement_pid)
        {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::OpenCodeServer, error.clone());
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        if let Err(error) = wait_for_runtime_client_process_readiness_with(
            &self.managed_process.spec,
            &mut replacement_child,
            clock,
            sleeper,
        ) {
            let replacement_child = Arc::new(Mutex::new(replacement_child));
            let mut failed_replacement_process = RunningRuntimeClientProcess {
                spec: self.managed_process.spec.clone(),
                child: replacement_child,
                output_capture: replacement_output_capture.clone(),
                platform_scope: self.managed_process.platform_scope.clone(),
            };
            let _ = stop_runtime_client_process(&mut failed_replacement_process, clock, sleeper);
            self.managed_process
                .supervisor_handle
                .mark_component_restarting(SupervisedComponent::OpenCodeServer, error.clone());
            self.managed_process
                .supervisor_handle
                .emit_component_healthcheck_failed(
                    SupervisedComponent::OpenCodeServer,
                    "restart_readiness_failed",
                    &error,
                    "readiness_http",
                    &[],
                );
            self.managed_process
                .restart_in_progress
                .store(false, Ordering::Relaxed);
            return Err(error);
        }
        {
            let mut child = match self.managed_process.child.lock() {
                Ok(child) => child,
                Err(error) => {
                    self.managed_process
                        .restart_in_progress
                        .store(false, Ordering::Relaxed);
                    return Err(format!("runtime client child lock was poisoned: {error}"));
                }
            };
            *child = replacement_child;
        }
        self.managed_process
            .supervisor_handle
            .replace_component_details(
                SupervisedComponent::OpenCodeServer,
                opencode_server_details_with_status(
                    &self.managed_process.spec,
                    Some(replacement_pid),
                    None,
                    "Alive",
                    "Ready",
                ),
            );
        self.managed_process
            .supervisor_handle
            .mark_component_healthy(SupervisedComponent::OpenCodeServer);
        self.managed_process
            .restart_in_progress
            .store(false, Ordering::Relaxed);
        Ok(())
    }
}

File: packages/sandboxd/process/control.go
package process

import (
	"fmt"
	"sync"
	"sync/atomic"

	"github.com/mistle/sandboxd/supervision"
	"github.com/mistle/sandboxd/timeutil"
)

type managedRuntimeClientProcess struct {
	mutex   sync.Mutex
	process *RunningRuntimeClientProcess
}

type processRestartFailurePhase string

const (
	processRestartSpawnFailure     processRestartFailurePhase = "spawn"
	processRestartReadinessFailure processRestartFailurePhase = "readiness"
)

func (managed *managedRuntimeClientProcess) Spec() RuntimeClientProcessSpec {
	managed.mutex.Lock()
	defer managed.mutex.Unlock()
	return managed.process.Spec
}

func (managed *managedRuntimeClientProcess) Stop(clock timeutil.Clock, sleeper timeutil.Sleeper) error {
	managed.mutex.Lock()
	defer managed.mutex.Unlock()
	return StopRuntimeClientProcess(managed.process, clock, sleeper)
}

func (managed *managedRuntimeClientProcess) PID() uint32 {
	managed.mutex.Lock()
	defer managed.mutex.Unlock()
	return managed.process.PID()
}

func (managed *managedRuntimeClientProcess) ExitObservation() (processExitObservation, error) {
	managed.mutex.Lock()
	defer managed.mutex.Unlock()

	processID := managed.process.PID()
	exited, err := ProcessHasExited(managed.process)
	if err != nil {
		return processExitObservation{}, err
	}
	if !exited {
		return processExitObservation{PID: processID}, nil
	}
	reason, fields := processExitEventFields(managed.process)
	return processExitObservation{
		PID:         processID,
		Exited:      true,
		Description: processExitDescription(managed.process),
		Reason:      reason,
		Fields:      fields,
	}, nil
}

func (managed *managedRuntimeClientProcess) Restart(
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
) (*RunningRuntimeClientProcess, processRestartFailurePhase, error) {
	managed.mutex.Lock()
	defer managed.mutex.Unlock()

	processSpec := managed.process.Spec
	_ = StopRuntimeClientProcess(managed.process, clock, sleeper)

	replacementProcess, err := StartRuntimeClientProcess(processSpec)
	if err != nil {
		return nil, processRestartSpawnFailure, err
	}
	if err := WaitForRuntimeClientProcessReadiness(replacementProcess, clock, sleeper); err != nil {
		_ = StopRuntimeClientProcess(replacementProcess, clock, sleeper)
		return nil, processRestartReadinessFailure, err
	}

	managed.process = replacementProcess
	return replacementProcess, "", nil
}

type processExitObservation struct {
	PID         uint32
	Exited      bool
	Description string
	Reason      string
	Fields      map[string]any
}

type CodexAppServerObservationHandle struct {
	mutex       sync.Mutex
	observation CodexAppServerObservation
}

func NewCodexAppServerObservationHandle(
	processSpec RuntimeClientProcessSpec,
	pid uint32,
	isAlive bool,
	lastExitStatus *string,
) *CodexAppServerObservationHandle {
	observation := CodexAppServerObservation{}
	UpdateCodexAppServerObservation(&observation, processSpec, &pid, isAlive, lastExitStatus)
	return &CodexAppServerObservationHandle{observation: observation}
}

func (handle *CodexAppServerObservationHandle) Snapshot() CodexAppServerObservation {
	handle.mutex.Lock()
	defer handle.mutex.Unlock()
	return handle.observation
}

func (handle *CodexAppServerObservationHandle) Replace(
	processSpec RuntimeClientProcessSpec,
	pid uint32,
	isAlive bool,
	lastExitStatus *string,
) error {
	handle.mutex.Lock()
	defer handle.mutex.Unlock()
	return UpdateCodexAppServerObservation(&handle.observation, processSpec, &pid, isAlive, lastExitStatus)
}

type CodexAppServerControlHandle struct {
	managedProcess managedCodexAppServerProcess
}

type OpenCodeServerControlHandle struct {
	managedProcess managedOpenCodeServerProcess
}

type managedCodexAppServerProcess struct {
	process           *managedRuntimeClientProcess
	observation       *CodexAppServerObservationHandle
	supervisorHandle  *supervision.SandboxdSupervisorHandle
	restartMutex      sync.Mutex
	restartInProgress atomic.Bool
}

type managedOpenCodeServerProcess struct {
	process           *managedRuntimeClientProcess
	supervisorHandle  *supervision.SandboxdSupervisorHandle
	restartMutex      sync.Mutex
	restartInProgress atomic.Bool
}

func (handle *CodexAppServerControlHandle) ObservationHandle() *CodexAppServerObservationHandle {
	return handle.managedProcess.observation
}

func (handle *CodexAppServerControlHandle) Restart(clock timeutil.Clock, sleeper timeutil.Sleeper) error {
	if err := validateProcessManagerTiming(clock, sleeper); err != nil {
		return err
	}
	handle.managedProcess.restartMutex.Lock()
	defer handle.managedProcess.restartMutex.Unlock()

	handle.managedProcess.restartInProgress.Store(true)
	defer handle.managedProcess.restartInProgress.Store(false)

	handle.managedProcess.supervisorHandle.MarkComponentStarting(supervision.ComponentCodexAppServer)
	replacementProcess, failurePhase, err := handle.managedProcess.process.Restart(clock, sleeper)
	if err != nil {
		reason, probeKind, eventErr := codexRestartFailureEvent(failurePhase)
		if eventErr != nil {
			return eventErr
		}
		markRestartFailure(
			handle.managedProcess.supervisorHandle,
			supervision.ComponentCodexAppServer,
			err,
			reason,
			probeKind,
		)
		return err
	}

	processID := replacementProcess.PID()
	if err := handle.managedProcess.observation.Replace(replacementProcess.Spec, processID, true, nil); err != nil {
		return err
	}
	handle.managedProcess.supervisorHandle.ReplaceComponentDetails(
		supervision.ComponentCodexAppServer,
		CodexAppServerDetailsWithStatus(replacementProcess.Spec, &processID, nil, "Alive", "Ready"),
	)
	handle.managedProcess.supervisorHandle.MarkComponentHealthy(supervision.ComponentCodexAppServer)
	return nil
}

func (handle *OpenCodeServerControlHandle) Restart(clock timeutil.Clock, sleeper timeutil.Sleeper) error {
	if err := validateProcessManagerTiming(clock, sleeper); err != nil {
		return err
	}
	handle.managedProcess.restartMutex.Lock()
	defer handle.managedProcess.restartMutex.Unlock()

	handle.managedProcess.restartInProgress.Store(true)
	defer handle.managedProcess.restartInProgress.Store(false)

	handle.managedProcess.supervisorHandle.MarkComponentStarting(supervision.ComponentOpenCodeServer)
	replacementProcess, failurePhase, err := handle.managedProcess.process.Restart(clock, sleeper)
	if err != nil {
		reason, probeKind, eventErr := openCodeRestartFailureEvent(failurePhase)
		if eventErr != nil {
			return eventErr
		}
		markRestartFailure(
			handle.managedProcess.supervisorHandle,
			supervision.ComponentOpenCodeServer,
			err,
			reason,
			probeKind,
		)
		return err
	}

	processID := replacementProcess.PID()
	handle.managedProcess.supervisorHandle.ReplaceComponentDetails(
		supervision.ComponentOpenCodeServer,
		OpenCodeServerDetailsWithStatus(replacementProcess.Spec, &processID, nil, "Alive", "Ready"),
	)
	handle.managedProcess.supervisorHandle.MarkComponentHealthy(supervision.ComponentOpenCodeServer)
	return nil
}

func markRestartFailure(
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	component supervision.SupervisedComponent,
	cause error,
	reason string,
	probeKind string,
) {
	errorText := cause.Error()
	supervisorHandle.MarkComponentRestarting(component, errorText)
	supervisorHandle.EmitComponentHealthcheckFailed(
		component,
		reason,
		errorText,
		probeKind,
		map[string]any{},
	)
}

func codexRestartFailureEvent(failurePhase processRestartFailurePhase) (string, string, error) {
	switch failurePhase {
	case processRestartSpawnFailure:
		return "restart_spawn_failed", "process_liveness", nil
	case processRestartReadinessFailure:
		return "restart_readiness_failed", "readiness_ws", nil
	default:
		return "", "", fmt.Errorf("unsupported Codex app-server restart failure phase: %s", failurePhase)
	}
}

func openCodeRestartFailureEvent(failurePhase processRestartFailurePhase) (string, string, error) {
	switch failurePhase {
	case processRestartSpawnFailure:
		return "restart_spawn_failed", "process_liveness", nil
	case processRestartReadinessFailure:
		return "restart_readiness_failed", "readiness_http", nil
	default:
		return "", "", fmt.Errorf("unsupported OpenCode server restart failure phase: %s", failurePhase)
	}
}
